emma-language/src/main/scala/org/emmalanguage/compiler/lang/opt/FoldForestFusion.scala

```diff
@@ -0,0 +1,828 @@
+
+            // Case 2: One child - fold to be cata-fused.
+            case Seq((child, subIndex)) => valIndex.get(root).fold(child -> subIndex)({
+              // Root denotes a comprehension of the following kind:
+              //
+              // val root = for {  // type DataBag[B]
+              //   x <- { ...; xs } // type DataBag[A]
+              //   $qs
+              // } yield $ret
+              //
+              // val child = root.fold(alg$j)
+              case cs.Comprehension(
+                (q @ cs.Generator(x, core.Let(Seq(), Seq(), core.Ref(xs)))) +:
+                qs,
+                cs.Head(ret)) if isLinear(q +: qs) =>
+
+                val core.DefCall(/* root */ _, DataBag.fold1, targs, Seq(Seq(alg$1))) = subIndex(child)
+
+                val A = x.info
+                val B = ret.tpe
+                val C = child.info
+
+                if (qs.isEmpty) {
+                  // Case 2.a: exactly one generator and no guards.
+                  //
+                  // val fun$Map$f = (x: A) => $ret
+                  // val alg$Map$f = alg.Map[A, B, C](fun$Map$f, alg$j)
+                  // val child = xs.fold[C](alg$Map$f)
+                  val mod = api.Sym[algebra.Map.type].asModule
+
+                  val fun = ret match {
+                    case core.Let(
+                      Seq(core.ValDef(s, core.DefCall(Some(core.Ref(f)), m, Seq(), Seq(Seq(core.Ref(`x`)))))),
+                      Seq(),
+                      core.Ref(t)) if s == t && m.name == api.TermName.app =>
+                      // $ret = { val s = f.apply(x); s }
+                      // call `f` directly
+                      f -> u.EmptyTree
+                    case _ =>
+                      // val fun$Map$f = (x: A) => $ret
+                      val nme = freshFun(mod)
+                      val tpe = api.Type.fun(Seq(A), B)
+                      val rhs = api.Lambda(Seq(x), ret)
+                      val lhs = api.ValSym(owner, nme, tpe)
+                      lhs -> rhs
+                  }
+
+                  val alg = { // val alg$Map$f = alg.Map[A, B, C](fun$Map$f, alg$j)
+                    val cls = mod.companion.asClass
+                    val tpe = api.Type(cls.toTypeConstructor, Seq(A, B, C))
+                    val rhs = core.DefCall(tgtOf(mod), appOf(mod), Seq(A, B, C), Seq(Seq(refOf(fun), alg$1)))
+                    val lhs = api.ValSym(owner, freshAlg(mod), tpe)
+                    lhs -> rhs
+                  }
+
+                  val fld = { // val child = xs.fold[C](alg$Map$f)
+                    val rhs = foldAlg(tgtOf(xs), child.info, lhsOf(alg))
+                    child -> rhs
+                  }
+
+                  child -> (subIndex ++ Seq(fun, alg, fld).filterNot(rhsOf(_) == u.EmptyTree))
+                } else if (returns(ret, x) && qs.forall({ case cs.Guard(_) => true })) {
+                  // Case 2.b: exactly one generator, $ret matching `{ x }`, and at least one guard
```

Comment:
I guess this doesn't cover the case of `xs.withFilter(p).map(f)`, but we can add this later if it's necessary.